onnxruntime/wasm/api.cc

```diff
@@ -387,3 +395,109 @@ char* OrtEndProfiling(ort_session_handle_t session) {
+  return (CHECK_TRAINING_STATUS(CreateTrainingSessionFromArray, g_env, options,
+                                training_checkpoint_state_handle, train_model, train_size,
+                                eval_model, eval_size, optimizer_model, optimizer_size,
+                                &training_session) == ORT_OK)
+             ? training_session
+             : nullptr;
+}
+
+int EMSCRIPTEN_KEEPALIVE OrtTrainingLazyResetGrad(ort_training_session_handle_t training_handle) {
+  return CHECK_TRAINING_STATUS(LazyResetGrad, training_handle);
+}
+
+OrtValue* EMSCRIPTEN_KEEPALIVE OrtRunTrainStep(ort_training_session_handle_t training_handle,
+                                               const ort_run_options_handle_t options,
+                                               const size_t inputs_len,
+                                               const ort_tensor_handle_t* inputs,
+                                               const size_t outputs_len,
+                                               ort_tensor_handle_t* outputs) {
+  return (CHECK_TRAINING_STATUS(TrainStep, training_handle,
+                                options, inputs_len, inputs, outputs_len, outputs) == ORT_OK)
+             ? *outputs
+             : nullptr;
+}
+
+int EMSCRIPTEN_KEEPALIVE OrtTrainingOptimizerStep(ort_training_session_handle_t training_handle,
+                                                  const ort_run_options_handle_t run_options) {
+  return CHECK_TRAINING_STATUS(OptimizerStep, training_handle, run_options);
+}
+
+ort_tensor_handle_t EMSCRIPTEN_KEEPALIVE OrtTrainingEvalStep(ort_training_session_handle_t training_handle,
+                                                             const ort_run_options_handle_t options,
+                                                             size_t inputs_len,
+                                                             const ort_tensor_handle_t* inputs,
+                                                             size_t outputs_len,
+                                                             ort_tensor_handle_t* outputs) {
+  return (CHECK_TRAINING_STATUS(EvalStep, training_handle,
+                                options, inputs_len, inputs, outputs_len, outputs) == ORT_OK)
+             ? *outputs
+             : nullptr;
+}
+
+// void EMSCRIPTEN_KEEPALIVE OrtTrainingExportModelForInferencing(orttraining_session_handle_t session,
+//                                                                const ORTCHAR_T* inference_model_path,
+//                                                                size_t graph_outputs_len,
+//                                                                const char* const* graph_output_names) {
+//   Ort::GetTrainingApi().ExportModelForInferencing(session, inference_model_path, graph_outputs_len, graph_output_names);
+// }
+
+size_t* EMSCRIPTEN_KEEPALIVE OrtTrainingGetParametersSize(ort_training_session_handle_t training_handle,
```

Comment:
can we not return size_t @fs-eire ?

Reply:
if we want to return a single value which cannot be zero ( 0 is consider illegal value), then we should just return it (when 0 indicating there is error).

if we want to return a single value which can be zero ( 0 is consider legal value), then we should accept a `__out size_t* p_value` in the function parameter list and return error code.

this implementation returns the address of a stack variable which is no longer valid after the returning of this function. this need to be fixed.